flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/fanout/FanOutProperties.java

```diff
@@ -0,0 +1,386 @@
+
+	/**
+	 * Exponential backoff power constant for the register stream operation.
+	 */
+	private final double registerStreamExpConstant;
+
+	/**
+	 * Maximum retry attempts for the register stream operation.
+	 */
+	private final int registerStreamMaxRetries;
+
+	/**
+	 * Base backoff millis for the deregister stream operation.
+	 */
+	private final long deregisterStreamBaseBackoffMillis;
+
+	/**
+	 * Maximum backoff millis for the deregister stream operation.
+	 */
+	private final long deregisterStreamMaxBackoffMillis;
+
+	/**
+	 * Exponential backoff power constant for the deregister stream operation.
+	 */
+	private final double deregisterStreamExpConstant;
+
+	/**
+	 * Maximum retry attempts for the deregister stream operation.
+	 */
+	private final int deregisterStreamMaxRetries;
+
+	/**
+	 * Base backoff millis for the list stream operation.
+	 */
+	private final long listStreamConsumersBaseBackoffMillis;
+
+	/**
+	 * Maximum backoff millis for the list stream operation.
+	 */
+	private final long listStreamConsumersMaxBackoffMillis;
+
+	/**
+	 * Exponential backoff power constant for the list stream operation.
+	 */
+	private final double listStreamConsumersExpConstant;
+
+	/**
+	 * Maximum retry attempts for the list stream operation.
+	 */
+	private final int listStreamConsumersMaxRetries;
+
+	// ------------------------------------------------------------------------
+	//  registerStream() related performance settings
+	// ------------------------------------------------------------------------
+	/**
+	 * Creates a FanOutProperties.
+	 *
+	 * @param configProps the configuration properties from config file.
+	 * @param streams     the streams which is sent to match the EFO consumer arn if the EFO registration mode is set to `NONE`.
+	 */
+	public FanOutProperties(Properties configProps, List<String> streams) {
+		Preconditions.checkArgument(configProps.getProperty(ConsumerConfigConstants.RECORD_PUBLISHER_TYPE).equals(RecordPublisherType.EFO.toString()), "Only efo record publisher can register a FanOutProperties.");
+		KinesisConfigUtil.validateEfoConfiguration(configProps, streams);
+
+		efoRegistrationType = EFORegistrationType.valueOf(configProps.getProperty(ConsumerConfigConstants.EFO_REGISTRATION_TYPE, EFORegistrationType.EAGER.toString()));
+		//if efo registration type is EAGER|LAZY, then user should explicitly provide a consumer name for each stream.
+		if (efoRegistrationType == EFORegistrationType.EAGER || efoRegistrationType == EFORegistrationType.LAZY) {
+			consumerName = configProps.getProperty(ConsumerConfigConstants.EFO_CONSUMER_NAME);
+		} else {
+			//else users should explicitly provide consumer arns.
+			streamConsumerArns = new HashMap<>();
+			for (String stream : streams) {
+				String key = ConsumerConfigConstants.EFO_CONSUMER_ARN_PREFIX + "." + stream;
+				streamConsumerArns.put(stream, configProps.getProperty(key));
+			}
+		}
+
+		this.subscribeToShardMaxRetries = Integer.parseInt(
```

Comment:
@xiaolong-sn nit: personally I would break these blocks into private methods for readability, but this is developer preference